src/lib.rs

```diff
@@ -11,6 +11,678 @@
+        config: MqttConfig,
+        client_id: UUID,
+        authority_name: String,
+        client_type: UPClientMqttType,
+    ) -> Result<Self, UStatus> {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let ptr = topic_listener_map.clone();
+
+        let mqtt_protocol = if config.ssl_options.is_some() {
+            "mqtts"
+        } else {
+            "mqtt"
+        };
+
+        let mqtt_uri = format!(
+            "{}://{}:{}",
+            mqtt_protocol, config.mqtt_hostname, config.mqtt_port
+        );
+
+        let mqtt_cli = mqtt::CreateOptionsBuilder::new()
+            .server_uri(mqtt_uri)
+            .client_id(client_id)
+            .max_buffered_messages(100)
+            .create_client()
+            .map_err(|e| {
+                UStatus::fail_with_code(
+                    UCode::INTERNAL,
+                    format!("Unable to create mqtt client: {e:?}"),
+                )
+            })?;
+
+        mqtt_cli.set_message_callback(
+            move |cli: &AsyncClient, message: std::option::Option<paho_mqtt::Message>| {
+                UPClientMqtt::on_receive(cli, message, ptr.clone())
+            },
+        );
+
+        // TODO: Integrate ssl options when connecting, may need a username, etc.
+        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
+            .clean_start(false)
+            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
+            .finalize();
+
+        mqtt_cli.connect(conn_opts).await.map_err(|e| {
+            UStatus::fail_with_code(
+                UCode::INTERNAL,
+                format!("Unable to connect to mqtt broker: {e:?}"),
+            )
+        })?;
+
+        Ok(Self {
+            mqtt_client: Arc::new(mqtt_cli),
+            topic_listener_map,
+            authority_name,
+            client_type,
+        })
+    }
+
+    /// Helper function that handles MQTT messages on reception.
+    ///
+    /// # Arguments
+    /// * `cli` - MQTT client that received the message.
+    /// * `message` - MQTT message received.
+    fn on_receive(
+        _cli: &AsyncClient,
+        message: Option<paho_mqtt::Message>,
+        topic_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    ) {
+        if let Some(msg) = message {
+            let topic = msg.topic();
+
+            // Get attributes from mqtt header.
+            let uattributes =
+                UPClientMqtt::get_uattributes_from_mqtt_properties(msg.properties()).unwrap();
+
+            let payload = msg.payload();
+            let upayload = payload.to_vec();
+
+            let umessage = UMessage {
+                attributes: Some(uattributes).into(),
+                payload: Some(upayload.into()),
+                ..Default::default()
+            };
+
+            // Create UMessage from UAttributes and UPayload.
+            let topic_map_read = block_on(topic_map.read());
```

Comment:
Do we need block_on here? Can't we .await?

Reply:
It is a good question, the challenge I am running into is this is a function that gets called every time the mqtt_client receives a message to a topic, and with paho_mqtt I have been unable to get that function/closure be async. This is used at line 117 `mqtt_cli.set_message_callback`. I can double check to see if there is a way to have that function be async

Reply:
So the AsyncClient.set_message_callback fn looks like

`pub fn set_message_callback<F>(&self, cb: F)
where
    F: FnMut(&AsyncClient, Option<Message>) + Send + 'static,`
    
Since async fn's send a Future as output, I do not believe I would be able to get around passing in an async function so I believe block_ons are needed

Reply:
I think the block_on here 
block_on(listener.on_receive(umsg_clone));
is more critical since it blocks on the mqtt library receiver thread. Basically if you have multiple listeners a slow listener will cause problems. I think a better way to do it, which would also fix the other issues would be to use a channel process the messages in the context of another thread (https://github.com/eclipse/paho.mqtt.rust/issues/5). It would also fix the current issue that you cannot do for example a tokio thread sleep in a listener because you are not on a tokio thread.

Reply:
Updated to pass messages over a channel to be handled by a tokio spawned task.